forcedotcom/salesforcedx-vscode
# Restructure and Improvement of E2E Tests Workflow

### What does this PR do?
This PR contains a restructure of the current E2E workflow on Github Actions to improve its usability, performance, and scalability.

### What issues does this PR fix or reference?
@W-13995839@

### Functionality Before

**How to Run:**
The E2E tests are run from the "End to End Tests" workflow in Github Actions.  To select which E2E test(s) to run, the user uses the specFiles dropdown and chooses the name of the test that they want to run.  _Users can run one test at a time, unless they choose "*.e2e.ts" to run all available E2E tests_.

**How to Rerun Failed Tests:**
Each run of the "End to End Tests" workflow produces one job.  Users can rerun the failed test(s) using the "Re-run jobs" button in the top right corner of the page.  If the user originally selected the "*.e2e.ts" option, _the "Re-run jobs" button will rerun all tests for that OS, including those that passed_.

**Performance When Multiple Tests Are Run:**
_All tests are run sequentially, in alphabetical order._  There are three VMs created: one for each OS.  Each VM does the setup once and runs all tests one after the other in the "Run Headless Test" step.  Sometimes (but not always) the failure of one E2E test causes subsequent tests to fail.

### Functionality After

**Changes Made:**
_Modified File:_
    • e2e.yml - Calls apexE2E.yml, coreE2E.yml, and/or lspE2E.yml based on which category(ies) are selected.
_New Files:_
    • apexE2E.yml - Contains all tests related to the Apex extensions. Calls runE2ETest.yml based on which E2E test(s) are selected.
    • coreE2E.yml - Contains all tests related to the CLI extension. Calls runE2ETest.yml based on which E2E test(s) are selected.
    • lspE2E.yml - Contains all tests related to LSPs except ApexLSP. Calls runE2ETest.yml based on which E2E test(s) are selected.
    • runE2ETest.yml - Creates and sets up a VM to run one E2E test that is passed in as an input.

**How to Run:**
The E2E tests are broken down into the Apex, Core, and LSP categories and placed under three separate workflows.  From each of the three workflows - (1) Apex End to End Tests (2) Core End to End Tests (3) LSP End to End Tests - users can use checkboxes to select one or more tests that they would like to run.  If the user wants to run all the tests in one or more categories, they can go to the "End to End Tests" workflow and check the box of the category(ies) that they want to run.

Examples:
    1. Run all tests in the Apex and Core categories: Go to the "End to End Tests" workflow and check the boxes for "Apex e2e Tests" and "Core e2e Tests".
    2. Run only "deployAndRetrieve.e2e.ts" and "templates.e2e.ts": Go to the "Core End to End Tests" workflow and check the boxes for "deployAndRetrieve.e2e.ts" and "templates.e2e.ts".

**How to Rerun Failed Tests:**
The user should go to the workflow where the job is initiated and rerun the failed test(s) using the "Re-run jobs" button in the top right corner of the page.  New VMs will be created to run the failed tests in parallel.  _Only the test(s) that fail will be rerun._

**Performance When Multiple Tests Are Run:**
_All E2E tests are run in parallel._  Each E2E test is run on three VMs (one for each OS).  _Each VM will do its own setup (clone, npm install, etc.) and then run one test._  Tests are independent and will not be affected by runs of other tests.  Based on current limitations of resource provisioning in Github Actions, if multiple categories are selected from the "End to End Tests" workflow, jobs might queue up until another job completes and a VM is available to run the queued job.
